File: Cloud Storage Samples/Dropbox Browser/c++/dropbox.cpp
/*
 * Cloud Storage 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#define LINE_LEN 100

class MyDropbox : public Dropbox
{
public:
	
	virtual int FireSSLServerAuthentication(DropboxSSLServerAuthenticationEventParams *e)
	{
		e->Accept = true;
		return 0;
	}
};

MyDropbox dropbox1;

void printOptions()
{
	printf( "Type ? to select this menu or a number 1-5 to perform one of the following actions:\n"
	        "1.  List Documents\n"
	        "2.  Download Document\n"
	        "3.  Upload Document\n"
	        "4.  Delete Document\n"
	        "5.  Quit\n");
}

char* resolveResourceType(int type) {
  switch (type)
  {
    case BRT_FILE: return "File";
    case BRT_FOLDER: return "Folder";
    default: return "Unknown";
  }
}

void displayDocuments(){

	int ret_code = dropbox1.ListResources("");
	if(ret_code)
		goto error;

	printf("%-3.1s%-40.38s%-10.8s%-25.23s\n\n","#","Title","Type","Modified Date");

	for(int i=0;i<dropbox1.GetResourceCount();i++)
	{
		printf("%-3.1d%-40.38s%-10.8s%-25.23s\n",i,dropbox1.GetResourceName(i),resolveResourceType(dropbox1.GetResourceType(i)),dropbox1.GetResourceModifiedTime(i));
	}
	return;

error:
	printf("\nError: %d", ret_code);
	if (dropbox1.GetLastError())
	{
		printf( " \"%s\"\n", dropbox1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	exit(ret_code);
}

int main(int argc, char **argv)
{	
  char buffer[LINE_LEN+1];
  int ret_code = 0;

  printf("***************************************************************\n");
  printf("* This demo shows how to use the Dropbox component to list,    *\n");
  printf("* download, upload, and delete documents. To begin, you will   *\n");
  printf("* first need to setup an app in the dropbox developer console  *\n");
  printf("* to obtain your Client Id and Client Secret. You will also    *\n");
  printf("* need to setup a Redirect URI to http://localhost:PORT, where *\n");
  printf("* PORT is the port number configured in the code.              *\n");
  printf("***************************************************************\n\n");

  printf("Client Id: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  dropbox1.SetOAuthClientId(buffer); 
    
  printf("Client Secret: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  dropbox1.SetOAuthClientSecret(buffer);
  
  dropbox1.SetOAuthServerAuthURL("https://www.dropbox.com/oauth2/authorize");
  dropbox1.SetOAuthServerTokenURL("https://api.dropboxapi.com/oauth2/token");
  dropbox1.Config("OAuthWebServerPort=7777");

  printf("\nPerforming OAuth authentication ...\n");

  dropbox1.Authorize();

  ret_code = dropbox1.GetLastErrorCode();

  if(ret_code)
    goto error;

	printf("\n");
	printOptions();

	while (1)
	{
		printf( "\n> " );
		scanf("%80s",buffer);

		if ( ! strcmp(buffer, "?") )
		{
			printOptions();
		}

		else if ( ! strcmp(buffer, "1") ) //List documents
		{
			displayDocuments();
		}

		else if ( ! strcmp(buffer, "2") ) //Download document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);
			int docNum = atoi(buffer);

			printf("Local Download Directory: ");
			scanf("%80s",buffer);
			strcat(buffer,"\\");
			strcat(buffer,dropbox1.GetResourceName(docNum));
			dropbox1.SetLocalFile(buffer);

			ret_code = dropbox1.DownloadFile(dropbox1.GetResourceId(docNum)); //Use the default file format
			if(ret_code)
				goto error;

			printf("\n Download Successful\n");
		}

		else if ( ! strcmp(buffer, "3") ) //Upload document
		{
			printf("\nLocal File: ");
			scanf("%80s",buffer);

			dropbox1.SetLocalFile(buffer);
			
			printf("Name of New File on Server: ");
			scanf("%80s",buffer);
      dropbox1.UploadFile(buffer);
			ret_code = dropbox1.GetLastErrorCode();

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "4") ) //Delete Document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);
			char* fileToDelete = dropbox1.GetResourceId(atoi(buffer));

			ret_code = dropbox1.DeleteResource(fileToDelete);

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "5") )
		{
			exit(0);
		}

		else if ( ! strcmp(buffer, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Please select a number from [1-5].\n" );
		} // end of command checking
	}  // end of main while loop
	return ret_code;

error:
	printf("\nError: %d", ret_code);
	if (dropbox1.GetLastErrorCode())
	{
		printf( " \"%s\"\n", dropbox1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	return ret_code;

}

File: Cloud Storage Samples/Dropbox Browser/net/dropbox.cs
/*
 * Cloud Storage 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using System.IO;
using nsoftware.CloudStorage;

class dropboxDemo
{
  private static Dropbox dropbox = new nsoftware.CloudStorage.Dropbox();

  static void Main(string[] args)
  {
    try
    {
      dropbox.OnSSLServerAuthentication += dropbox_OnSSLServerAuthentication;

      // Prompt for authentication information.
      Console.Write("Enter your OAuth client ID: ");
      dropbox.OAuth.ClientId = Console.ReadLine();
      Console.Write("Enter your OAuth client secret: ");
      dropbox.OAuth.ClientSecret = Console.ReadLine();

      dropbox.OAuth.ServerAuthURL = "https://www.dropbox.com/oauth2/authorize";
      dropbox.OAuth.ServerTokenURL = "https://api.dropboxapi.com/oauth2/token";

      Console.Write("Enter a redirect URI port: ");
      dropbox.Config("OAuthWebServerPort=" + Console.ReadLine());
      dropbox.Config("OAuthBrowserResponseTimeout=60");
      dropbox.Authorize();

      // Process user commands.
      Console.WriteLine("Type \"?\" or \"help\" for a list of commands.");
      Console.Write("dropbox> ");
      string command;
      string[] arguments;

      while (true)
      {
        command = Console.ReadLine();
        arguments = command.Split();

        if (arguments[0] == "?" || arguments[0] == "help")
        {
          Console.WriteLine("Commands: ");
          Console.WriteLine("  ?                            display the list of valid commands");
          Console.WriteLine("  help                         display the list of valid commands");
          Console.WriteLine("  ls                           list resources in the root folder");
          Console.WriteLine("  get <resource number>        download the file resource with the specified number from the list");
          Console.WriteLine("  put <file>                   upload a new file resource");
          Console.WriteLine("  del <resource number>        delete the resource with the specified number from the list");
          Console.WriteLine("  quit                         exit the application");
        }
        else if (arguments[0] == "ls")
        {
          ListDocuments();
        }
        else if (arguments[0] == "get")
        {
          if (arguments.Length > 1)
          {
            DropboxResource resource = dropbox.Resources[int.Parse(arguments[1])];
            dropbox.LocalFile = resource.Name;
            dropbox.DownloadFile(resource.Id); // Use the default file format.

            Console.WriteLine("Download complete.");
          }
        }
        else if (arguments[0] == "put")
        {
          if (arguments.Length > 1)
          {
            dropbox.LocalFile = arguments[1];
            FileInfo file = new FileInfo(dropbox.LocalFile);
            dropbox.UploadFile(file.Name);

            Console.WriteLine("Upload complete.");
          }
        }
        else if (arguments[0] == "del")
        {
          if (arguments.Length > 1)
          {
            DropboxResource resource = dropbox.Resources[int.Parse(arguments[1])];
            dropbox.DeleteResource(resource.Id);

            Console.WriteLine("Deletion complete.");
          }
        }
        else if (arguments[0] == "quit")
        {
          break;
        }
        else if (arguments[0] == "")
        {
          // Do nothing.
        }
        else
        {
          Console.WriteLine("Invalid command.");
        } // End of command checking.

        Console.Write("dropbox> ");
      }
    }
    catch (Exception ex)
    {
      Console.WriteLine(ex.Message);
    }
  }

  private static void ListDocuments()
  {
    try
    {
      dropbox.ListResources("");

      Console.WriteLine();
      Console.WriteLine("{0,-3}{1,-40}{2,-20}{3,-30}", "#", "Title", "Type", "Last Modified");
      Console.WriteLine("--------------------------------------------------------------------------------");

      for (int i = 0; i < dropbox.Resources.Count; i++)
      {
        Console.WriteLine("{0,-3}{1,-40}{2,-20}{3,-30}", i, dropbox.Resources[i].Name, ResolveResourceType(dropbox.Resources[i].Type), dropbox.Resources[i].ModifiedTime);
      }
      Console.WriteLine();
    }
    catch (Exception ex)
    {
      Console.WriteLine(ex.Message);
    }
  }

  private static string ResolveResourceType(DropboxResourceTypes type)
  {
    switch (type)
    {
      case DropboxResourceTypes.drtFile:
        return "File";
      case DropboxResourceTypes.drtFolder:
        return "Folder";
      default:
        return "Unknown";
    }
  }

  private static void dropbox_OnSSLServerAuthentication(object sender, DropboxSSLServerAuthenticationEventArgs e)
  {
    if (e.Accept) return;
    Console.Write("Server provided the following certificate:\nIssuer: " + e.CertIssuer + "\nSubject: " + e.CertSubject + "\n");
    Console.Write("The following problems have been determined for this certificate: " + e.Status + "\n");
    Console.Write("Would you like to continue anyways? [y/n] ");
    if (Console.Read() == 'y') e.Accept = true;
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add a key to the dictionary for each argument.
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/", then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}